pgrx/src/memcx.rs

```diff
@@ -45,3 +55,99 @@ where
+}
+
+#[cfg(feature = "nightly")]
+unsafe impl<'mcx> std::alloc::Allocator for &MemCx<'mcx> {
+    // Future-proofing - currently only pg16 supports arbitrary alignment, but that is likely
+    // to change, whereas old versions are unlikely to lose 8-byte alignment.
+    #[cfg(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15"))]
+    fn allocate(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
+        // On versions before Postgres 16, alignment is always 8 byte / 64 bit.
+        let layout = layout_for_pre16(layout)?;
+        unsafe {
+            let ptr = pgrx_pg_sys::MemoryContextAlloc(self.ptr.as_ptr(), layout.size());
+            let slice: &mut [u8] = slice::from_raw_parts_mut(ptr.cast(), layout.size());
+            Ok(NonNull::new_unchecked(slice))
+        }
+    }
+
+    #[cfg(not(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15")))]
+    fn allocate(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
+        // Postgres 16 and newer permit any arbitrary power-of-2 alignment
+        unsafe {
+            // Bitflags for MemoryContextAllocAligned:
+            // #define MCXT_ALLOC_HUGE    0x01 /* allow huge allocation (> 1 GB) */
+            // #define MCXT_ALLOC_NO_OOM  0x02 /* no failure if out-of-memory */
+            // #define MCXT_ALLOC_ZERO    0x04 /* zero allocated memory */
+            let ptr = pgrx_pg_sys::MemoryContextAllocAligned(
+                self.ptr.as_ptr(),
+                layout.size(),
+                layout.align(),
+                0,
+            );
+            let slice: &mut [u8] = slice::from_raw_parts_mut(ptr.cast(), layout.size());
+            Ok(NonNull::new_unchecked(slice))
+        }
+    }
+
+    unsafe fn deallocate(&self, ptr: NonNull<u8>, _layout: std::alloc::Layout) {
+        // TODO: Find faster free for use when MemoryContext is known.
+        // This is the global function that looks up the relevant Memory Context by address range.
+        pgrx_pg_sys::pfree(ptr.as_ptr().cast())
+    }
+
+    #[cfg(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15"))]
+    fn allocate_zeroed(
+        &self,
+        layout: std::alloc::Layout,
+    ) -> Result<NonNull<[u8]>, std::alloc::AllocError> {
+        // Overriding default function here to use Postgres' zeroing implementation.
+        // On versions before Postgres 16, alignment is always 8 byte / 64 bit.
+        let layout = layout_for_pre16(layout)?;
+        unsafe {
+            let ptr = pgrx_pg_sys::MemoryContextAllocZero(self.ptr.as_ptr(), layout.size());
+            let slice: &mut [u8] = slice::from_raw_parts_mut(ptr.cast(), layout.size());
+            Ok(NonNull::new_unchecked(slice))
+        }
+    }
+
+    #[cfg(not(any(feature = "pg12", feature = "pg13", feature = "pg14", feature = "pg15")))]
```

Comment:
I don't know if @workingjubilee has a preference, but writing this as just `#[cfg(feature = "pg16")]` would be a lot easer to reason about.

Reply:
Corrected in https://github.com/pgcentralfoundation/pgrx/pull/1737/commits/f149072c318f7e39c212daf3b59e0f723d1dfd46